core/src/main/kotlin/fr/sncf/osrd/stdcm/infra_exploration/StepTracker.kt

```diff
@@ -0,0 +1,118 @@
+package fr.sncf.osrd.stdcm.infra_exploration
+
+import fr.sncf.osrd.graph.PathfindingEdgeLocationId
+import fr.sncf.osrd.sim_infra.api.Block
+import fr.sncf.osrd.sim_infra.api.BlockId
+import fr.sncf.osrd.sim_infra.api.TravelledPath
+import fr.sncf.osrd.stdcm.STDCMStep
+import fr.sncf.osrd.utils.units.Offset
+
+/**
+ * Component of `InfraExplorer` in charge of keeping track of anything related to the input steps.
+ * It keeps tracks of which steps we've passed and where, how many steps have been reached
+ * (specifically when we've reached the destination), and which steps are present in a block range.
+ *
+ * The tricky part is that we need to keep track of which steps are seen on the path as a whole
+ * (including lookahead), and which have actually been reached by the simulations (excluding
+ * lookahead). Unless specified otherwise, fields and methods refer to the whole path.
+ */
+class StepTracker(private val inputSteps: List<STDCMStep>) {
+    // If copying this is too expensive, it may be changed to an AppendOnlyLinkedList.
+    // But the list should be small.
+    private val reachedSteps: MutableList<LocatedStep> = mutableListOf()
+    private val stepsIncludingLookahead: Int
+        get() = reachedSteps.size
+
+    var stepsExcludingLookahead: Int = 0
+        private set
+
+    // Used to compute path offsets
+    private var currentPathOffset: Offset<TravelledPath> = Offset.zero()
+
+    /** Returns all the steps that have been passed on the path, in order. */
+    fun getAllReachedSteps(): List<LocatedStep> {
+        return reachedSteps
+    }
+
+    /** True if the last step has been reached (including lookahead). */
+    fun hasReachedDestination(): Boolean {
+        return stepsIncludingLookahead == inputSteps.size
+    }
+
+    /** True if the last step has been reached, with full simulation and no lookahead. */
+    fun hasReachedDestinationExcludingLookahead(): Boolean {
+        return hasReachedDestination() && stepsIncludingLookahead == stepsExcludingLookahead
```

Comment:
It would be clearer if the condition was the same as the above method no?
```suggestion
        return hasReachedDestination() && stepsExcludingLookahead == inputSteps.size
```

Reply:
Fixed [here](https://github.com/OpenRailAssociation/osrd/pull/11109/commits/09b5967ef81f6ab0b026d5dbe1e0201d2b6a95ed#diff-cf16bc37931f8039dd7a63ad658eaf552824657169fb8595f9d695f88e9a151aR45-R47)